learn: communicate with JS (#12)
melange-re/documentation-site: docs/learn.md

## Communicate with JavaScript

Melange offers an extensive range of techniques to use JavaScript code in
Melange projects and to make Melange code accessible to JavaScript programs. To
learn about these techniques, we will go through the types in Melange and how

> they map to JavaScript runtime code, then we will see what are the OCaml

Author: I took this suggestion and adapted it to surrounding text in https://github.com/melange-re/documentation-site/pull/12/commits/77b246e1cb80d752e0f0c2b1f297a8484cb54847.

language extensions that make them possible, and also introduce the `external`
construct. Finally, we will provide a variety of use cases with examples to show
how to communicate to and from JavaScript.

### Data types and runtime representation

This is how each Melange type is converted into JavaScript values:

Melange | JavaScript
---------------------|---------------

> int | number

Author: Do you think any further note would be useful here?

> int | number

Reviewer: probably not. good as is.

nativeint | number
int32 | number
float | number
string | string
array | array
tuple `(3, 4)` | array `[3, 4]`
bool | boolean
Js.Nullable.t | `null` / `undefined`
Option.t `None` | `undefined`
Option.t `Some( Some .. Some (None))` | internal representation
Option.t `Some 2` | `2`
record `{x: 1; y: 2}` | object `{x: 1, y: 2}`
int64 | array of 2 elements `[high, low]` high is signed, low unsigned
char | `'a'` -> `97` (ascii code)
bytes | number array
list `[]` | `0`

> list `[x, y]` | `[x, [y, 0]]`

Author: Fixed in https://github.com/melange-re/documentation-site/pull/12/commits/c9fa4fe982809950f537bca9ffd590e1c0d78295 (also the line below).

list `[1, 2, 3]` | `[ 1, [ 2, [ 3, 0 ] ] ]`
variant | See below
polymorphic variant | See below

Variants with a single non-nullary constructor:

```ocaml
type tree = Leaf | Node of int * tree * tree
(* Leaf -> 0 *)
(* Node(7, Leaf, Leaf) -> { _0: 7, _1: 0, _2: 0 } *)
```

Variants with more than one non-nullary constructor:

```ocaml
type t = A of string | B of int
(* A("foo") -> { TAG: 0, _0: "Foo" } *)
(* B(2) -> { TAG: 1, _0: 2 } *)
```

Polymorphic Variant:

```ocaml
let u = `Foo (* "Foo" *)
let v = `Foo(2) (* { NAME: "Foo", VAL: "2" } *)
```

> > **_NOTE:_** Relying on runtime representations by reading or writing them

Author: > but sometimes it's necessary. We shouldn't be so stern here.

Would you remove the note altogether, or reword it? The intent of the note was to made clear that it is possible, but dangerous from a typing perspective (which I think it is, same as writing bindings).

> Perhaps we need to document which representations are stable and which ones aren't.

This is documented already:

https://github.com/melange-re/documentation-site/blob/c9fa4fe982809950f537bca9ffd590e1c0d78295/docs/learn.md?plain=1#L676

https://github.com/melange-re/documentation-site/blob/c9fa4fe982809950f537bca9ffd590e1c0d78295/docs/learn.md?plain=1#L757

> > **_NOTE:_** Relying on runtime representations by reading or writing them

Reviewer: ok then nothing actionable?

> > **_NOTE:_** Relying on runtime representations by reading or writing them

Author: I tried to tackle your original comment by rewriting the paragraph and removing parts of it in https://github.com/melange-re/documentation-site/pull/12/commits/b2e6ef741b09c1053470b0626d40b3358d6b49ba.

> > **_NOTE:_** Relying on runtime representations by reading or writing them

Reviewer: nice, that reads a lot better.

> > **_NOTE:_** Relying on runtime representations by reading or writing them

Author: cool, resolving this conversation then (for the sake of tracking remaining threads)

manually from JavaScript code that communicates with Melange code might lead to
runtime errors, as these representations might be changed in the future. In
general, it is safer to leave the Melange compiler deal with the manipulation of
these runtime values and interact with them from Melange code using bindings.

Let’s see now some of these types in detail.

### Shared types

The following are types that can be shared between Melange and JavaScript almost
"as is". Specific caveats are mentioned on the sections where they apply.

#### Strings

> Melange strings are an immutable sequence of bytes. This is problematic when

Author: Reworded in https://github.com/melange-re/documentation-site/pull/12/commits/f41a0fa018ed50b2529a571e9ee466a6a8652447.

> Melange strings are an immutable sequence of bytes. This is problematic when

Author: resolving as agreed offline that the new changes make sense.

interacting with JavaScript code, because if one tries to use some unicode
characters, like:

```ocaml
let () = Js.log "你好"
```

It will lead to some cryptic console output. To rectify this, Melange allows to
define [quoted string
literals](https://v2.ocaml.org/manual/lex.html#sss:stringliterals) using the
`js` identifier, for example:

```ocaml
let () = Js.log {js|你好，
世界|js}
```

> For convenience, Melange exposes another special quoted string identifier: `j`.

Author: It is documented later on:

https://github.com/melange-re/documentation-site/blob/c9fa4fe982809950f537bca9ffd590e1c0d78295/docs/learn.md?plain=1#L1624-L1625

I think it's better to keep `{json||json}` in context, as it is a pretty niche feature. Wdyt?

> For convenience, Melange exposes another special quoted string identifier: `j`.

Reviewer: fine with me.

It is similar to JavaScript’ string interpolation, but for variables only (not
arbitrary expressions):

```ocaml
let world = {j|世界|j}
let helloWorld = {j|你好，$world|j}
```

You can surround the interpolation variable in parentheses too: `{j|你
好，$(world)|j}`.

To work with strings, Melange standard library provides some utilities in the
[`Stdlib.String` module](todo-fix-me.md). The bindings to the native JavaScript
functions to work with strings are in the [`Js.String` module](todo-fix-me.md).

#### Float

Melange floats are JavaScript numbers, and vice-versa. The Melange standard

> library doesn’t come with a Float module. The JavaScript Float API can be found

Author: https://github.com/melange-re/documentation-site/pull/12/commits/e6c23fc2ea7f4b223899f2ea648d13b2fb2e39eb

in the [`Js.Float`](todo-fix-me.md) module.

#### Int

> **Ints are 32-bits**. They compile to JavaScript numbers but treating both as

Author: I tried to adapt the text, but then a few lines below we mention:

> treating both as the same might lead to unexpected behavior, because of the differences in precision

as JS ints might get up to 2^53 - 1. So we would need to explain how JavaScript can adapt the integers depth dynamically based on the operations applied to them.

I fear it'd be too much detail to include these JS details for the more generic / entry level manual to Melange. Wdyt?

> **Ints are 32-bits**. They compile to JavaScript numbers but treating both as

Author: Actually, I changed my mind and ended up adding explanation about all this in https://github.com/melange-re/documentation-site/pull/12/commits/e6c23fc2ea7f4b223899f2ea648d13b2fb2e39eb. 🤷‍♂️

> **Ints are 32-bits**. They compile to JavaScript numbers but treating both as

Reviewer: > integers themselves have a [max depth of 53

this is wrong, after 2^32-1 they're floating point.

> **Ints are 32-bits**. They compile to JavaScript numbers but treating both as

Reviewer: the rest looks good though.

> **Ints are 32-bits**. They compile to JavaScript numbers but treating both as

Author: Gave another try in https://github.com/melange-re/documentation-site/pull/12/commits/1484685819554d08386a6ab357c482e7265bfee7, lmk what you think.

> **Ints are 32-bits**. They compile to JavaScript numbers but treating both as

Author: And https://github.com/melange-re/documentation-site/pull/12/commits/cc4423bfe041f07eddd8b9b0b7448c13b506296c.

the same might lead to unexpected behavior, because of the differences in
precision. When working with large numbers, it is better to use floats. For
example, the bindings in `Js.Date` use floats. The bindings to work with

> JavaScript integers are in the [`Js.Int`](todo-fix-me.md) module.

Author: Added in https://github.com/melange-re/documentation-site/pull/12/commits/e6c23fc2ea7f4b223899f2ea648d13b2fb2e39eb.

#### Array

> Idiomatic OCaml arrays are supposed to be fix-sized. This constraint is relaxed

Reviewer: there's still "fix-sized" instead of "fixed-size"

> Idiomatic OCaml arrays are supposed to be fix-sized. This constraint is relaxed

Author: Oops. Fixed in https://github.com/melange-re/documentation-site/pull/12/commits/630615e1bc3421d52107b40b7d8fdb0292488792.

> on the Melange side. You can change its length using the usual [JavaScript Array

Author: > is that actually true? 

Changing the length is possible. The linked issue only allowed to do so in a safe way, right?

I rewrote this section in https://github.com/melange-re/documentation-site/pull/12/commits/e6c23fc2ea7f4b223899f2ea648d13b2fb2e39eb, to include also homomorphism of Melange arrays as a difference against JavaScript arrays.

> on the Melange side. You can change its length using the usual [JavaScript Array

Author: https://github.com/melange-re/documentation-site/pull/12/commits/e6c23fc2ea7f4b223899f2ea648d13b2fb2e39eb

API](todo-fix-me.md). Melange’s functions to work with arrays are in the
[`Js.Array`](todo-fix-me.md) module.

#### Tuple

> OCaml tuples are compiled to JavaScript arrays. This is convenient when writing

Author: I added an example in https://github.com/melange-re/documentation-site/pull/12/commits/8980c2eccab3f91f0462a2be188964406127a725. Please let me know if it was what you had in mind.

> OCaml tuples are compiled to JavaScript arrays. This is convenient when writing

Reviewer: looks good.

  `Js.Nullable.toOption` functions in the [`Js.Nullable` module](todo-fix-me.md)
  to transform them into either `null` or `undefined` values.
- List (also a variant type): use `Array.of_list` and `Array.to_list` in the
  [`Array` module](todo-fix-me.md).
- Character
- Int64
- Lazy values

### Language extensions

In order to interact with JavaScript, Melange needs to extend the language to
provide blocks that express these interactions.

One approach could be to introduce new syntactic constructs (keywords and such)
to do so, for example:

```ocaml
javascript add : int -> int -> int = "function(x,y){
  return x + y
}
```

But this would break compatibility with OCaml, which is one of the main goals of
Melange.

Fortunately, OCaml provides mechanisms to extend its language without breaking
compatibility with the parser or the language. These mechanisms are composed by
two parts:
- First, some syntax additions to define parts of the code that will be extended
  or replaced
- Second, compile-time OCaml native programs called [PPX
  rewriters](https://ocaml.org/docs/metaprogramming), that will read the syntax
  additions defined above and proceed to extend or replace them

> The syntax additions come in two flavours, called [extensions

Author: https://github.com/melange-re/documentation-site/pull/12/commits/5117a37a47e517cb6b0cb06268f42771367eebfa

nodes](https://v2.ocaml.org/manual/extensionnodes.html) and
[attributes](https://v2.ocaml.org/manual/attributes.html).

Extension nodes are blocks that are supposed to be replaced by a specific type
of PPX rewriters called extenders. Extension nodes use the `%` character to be
identified. Extenders will take the extension node and replace it with valid
OCaml code.

An example where Melange uses extensions to communicate with JavaScript is to
produce "raw" JavaScript inside a Melange program: 

```ocaml
[%%bs.raw "var a = 1; var b = 2"]
let add = [%bs.raw "a + b"]
```

Which will generate the following JavaScript code:

```js
var a = 1; var b = 2
var add = a + b
```

The difference between one and two `%` characters is detailed in the [OCaml
documentation](https://v2.ocaml.org/manual/extensionnodes.html).

> Attributes are "lighter" than extensions, and are used to annotate specific

Author: And https://github.com/melange-re/documentation-site/pull/12/commits/8bd60df88f6c0fffc6470361bca39ef570e7aed2 😓

parts of the code. For example, it is possible to have fields in a record
renamed using using the `bs.as` attribute:

```ocaml
type t = {
  foo : int; [@bs.as "foo_for_js"]
  bar : string;
}

let t = { foo = 2; bar = "b" }
```

This will generate the following JavaScript code:

```js
var t = {
  foo_for_js: 2,
  bar: "b"
};
```

To learn more about preprocessors, attributes and extension nodes, check the
[section about PPX
rewriters](https://ocaml.org/docs/metaprogramming#ppx-rewriters) in the OCaml
docs.

### List of attributes and extensions

> **_NOTE:_** All these attributes and extensions are prefixed with `bs.` for
> backwards compatibility. They will be updated to `mel.` in the future.

**Attributes:**

Used to annotate `external` definitions:

- [`bs.get`](#bind-to-object-properties): read JavaScript object properties

>   statically by name

Author: I am not sure I understand what the suggestion is, could you elaborate?

>   statically by name

Author: I got it now, added a reference to "dot notation" following [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Property_accessors, and example in the corresponding section: https://github.com/melange-re/documentation-site/pull/12/commits/54cc220edc62adaa052f0295f2b3ee757d53eeea

>   statically by name

Author: Let me know if this is enough. Note this section idea was to be as brief as possible (mostly enumeration of the available attrs and extensions and 1-line sentence to define them).

>   statically by name

Reviewer: yeah, nice.

- [`bs.get_index`](#bind-to-object-properties): read a JavaScript object’s
  properties dynamically by using the bracket notation `[]`
- [`bs.module`](#using-functions-from-other-javascript-modules): bind to a value
  from a JavaScript module
- [`bs.new`](#javascript-classes): bind to a JavaScript constructor
- [`bs.obj`](#using-jst-objects): create JavaScript object
- [`bs.return`](#wrapping-returned-nullable-values): automate conversion from
  nullable values to `Option.t` values

> - [`bs.send`](#calling-a-function-as-an-object-property): call a function that

Author: Yes, much better. Added in https://github.com/melange-re/documentation-site/pull/12/commits/28907dc02a0e4cc37a2ddc6d2502db5c777c1dd8, plus tweaked the titles in `bs.send` section.

- `bs.as`: redefine the name generated in the JavaScript output code. Used in
  [constant function arguments](#constant-values-as-arguments),
  [variants](todo-fix-me.md), [polymorphic
  variants](#using-polymorphic-variants-to-bind-to-enums-and-string-types) and
  [record fields](#objects-with-static-shape-record-like).
- [`bs.deriving`](todo-fix-me.md): generate getters and setters for some types
- [`bs.inline`](todo-fix-me.md): forcefully inline constant values
- [`bs.optional`](todo-fix-me.md): omit fields in a record (combines with
  `bs.deriving`)

**Extensions:**

In order to use any of these extensions, you will have to add the melange PPX
preprocessor to your project. To do so, add the following to the `dune` file:

```bash
(library
 (name lib)
 (modes melange)
 (preprocess
   (pps melange.ppx)))
```

The same field `preprocess` can be added to `melange.emit`.

Here is the list of all extensions supported by Melange:

- `bs.debugger`: insert `debugger` statements
- `bs.external`: read global values
- `bs.obj`: create JavaScript object literals
- `bs.raw`: write raw JavaScript code
- `bs.re`: insert regular expressions

> ### Intro to `external`

Reviewer: should we call this section "foreign function interface" or "modeling foreign JavaScript code"?

Author: Agree. Renamed in https://github.com/melange-re/documentation-site/pull/12/commits/d3da08456c9e3cdddb02a84e1348633f764e342f.

will be further explain in the next sections.

Once declared, one can use an `external` as a normal value.

Melange external functions are turned into the expected JavaScript values,
inlined into their callers during compilation, and completely erased afterwards.
They don’t appear in the JavaScript output, so there are no costs on bundling
size.

**Note**: it is recommended to use external functions and the `[@bs.xxx]`
attributes in the interface files as well, as this allows some optimizations
where the resulting JavaScript values can be directly inlined at the call sites.

#### Special identity external

One external worth mentioning is the following one:

```ocaml
type foo = string
type bar = int
external danger_zone : foo -> bar = "%identity"
```

This is a final escape hatch which does nothing but convert from the type `foo`
to `bar`. In the following sections, if you ever fail to write an `external`,
you can fall back to using this one. But try not to.

Let’s now see all the ways to use JavaScript from Melange.

### Generate raw JavaScript

It is possible to directly write JavaScript code from a Melange file. This is

> unsafe, but it can be useful for prototyping or as a escape hatch.

Reviewer: ```suggestion
unsafe, but it can be useful for prototyping or as an escape hatch.
```

To do it, we will use the `bs.raw`
[extension](https://v2.ocaml.org/manual/extensionnodes.html):

```ocaml
let add = [%bs.raw {|
  function(a, b) {
    console.log("hello from raw JavaScript!");
    return a + b;
  }
|}]

let () = Js.log (add 1 2)
```

The `{||}` strings are called ["quoted
strings"](https://ocaml.org/manual/lex.html#quoted-string-id). They are similar
to JavaScript’s template literals, in the sense that they are multi-line, and

> there is no need to escape characters inside the string.

Reviewer: but they don't support interpolation.

Author: But some of the quoted strings can interpolate :P https://reasonml.github.io/en/try?rrjsx=true&reason=DYUwLgBJC8EEQDMD2S4G4BQHSQK4VgG8ArAHwAkRhgkIASMU4gXyA

I am not sure how to tackle this suggestion, I think it'd be better to leave it as is, and not explain any more that these two similarities (multi-line and escaping), but if you have any ideas on how to explain the interpolation limitation while explaining why `{j||j}` does support it, happy to add them.

Using one `%` (`[%bs.raw <string>]`) is useful to define expressions (function
bodies, or other values) where the implementation is directly JavaScript. This
is useful as we can attach the type signature already in the same line, to make
our code safer. For example:

```ocaml
[%%bs.raw "var a = 1"]

let f : unit -> int = [%bs.raw "function() {return 1}"]
```

> Using two percentages `%%` (`[%%bs.raw <string>]`) is reserved for top-level

Reviewer: ```suggestion
Using two percentage signs (`[%%bs.raw <string>]`) is reserved for top-level
```

Reviewer: not just top-level definitions, but rather structure items, e.g. 

```ocaml

module Foo = struct
  [%%bs.raw "not really toplevel"]
end
```

Author: I did not include this particular example for simplicity. But removed the "top-level" reference, mentioned structure and signature definitions, and linked to OCaml manual in https://github.com/melange-re/documentation-site/pull/12/commits/787e2619e9ca835178c61918f01b9ad45ed97eac.

definitions, for example:

```ocaml
[%%bs.raw "var a = 1"]
```

### Debugger

> Melange allows to inject a `[%bs.debugger]` expression:

Reviewer: ```suggestion
Melange allows to inject a `debugger;` expression:
```

```ocaml
let f x y =
  [%bs.debugger];
  x + y
```

Output:

```javascript
function f (x,y) {
  debugger; // JavaScript developer tools will set an breakpoint and stop here
  x + y;
}
```

### Detect global variables

Melange provides a relatively type safe approach to use globals that might be
defined either in the JavaScript runtime environment: `bs.external`.

`[%bs.external id]` will check if the JavaScript value `id` is `undefined` or
not, and return an `Option.t` value accordingly.

For example:

```ocaml
let () = match [%bs.external __DEV__] with
| Some _ -> Js.log "dev mode"
| None -> Js.log "production mode"
```

Another example:

```ocaml
let () = match [%bs.external __filename] with
| Some f -> Js.log f
| None -> Js.log "non-node environment"
```

### Bind to JavaScript objects

JavaScript objects are used in a variety of use cases:

> - As a fixed shape "record" or "struct" in other languages (like OCaml and C).

Reviewer: struct is a bit misleading here because of OCaml's `module X = struct .... end`. should we omit it?

Author: Yeah, removed in https://github.com/melange-re/documentation-site/pull/12/commits/3fb7df5d8288cee48739f79312e06acaf6a5e048.

(TIL that C# has [record structs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/record-structs)! #allthethings)

- As a map or dictionary.
- As a class.
- As a module to import/export.

> Melange cleanly separates the binding methods for JavaScript objects based on

Reviewer: cleanly seems unnecessary (compared to what?)
```suggestion
Melange separates the binding methods for JavaScript objects based on
```

Author: Agree, in general I have been trying to remove as many adjectives as possible, and stick to purely descriptive language.

Removed in https://github.com/melange-re/documentation-site/pull/12/commits/3fb7df5d8288cee48739f79312e06acaf6a5e048.

If you want or need to use different field names on the Melange and the
JavaScript sides, you can use the `bs.as` decorator:

```ocaml
type action = {
  type_ : string [@bs.as "type"]
}

let action = { type_ = "ADD_USER" }
```

Which generates the JavaScript code:

```js
var action = {
  type: "ADD_USER"
};
```

This is useful to map to JavaScript attribute names that cannot be expressed in
Melange, for example, where the JavaScript name we want to generate is a
[reserved keyword](https://v2.ocaml.org/manual/lex.html#sss:keywords).

It is also possible to map a Melange record to a JavaScript array by passing
indices to the `bs.as` decorator:

```ocaml
type t = {
  foo : int; [@bs.as "0"]
  bar : string; [@bs.as "1"]
}

let value = { foo = 7; bar = "baz" }
```

And its JavaScript generated code:

```js
var value = [
  7,
  "baz"
];
```

##### Using `Js.t` objects

Alternatively to records, Melange offers another type that can be used to
produce JavaScript objects. This type is `'a Js.t`, where `'a` is an [OCaml
object](https://v2.ocaml.org/manual/objectexamples.html).

The advantage of objects versus records is that no type declaration is needed in

> advanced, which can be helpful for prototyping or quickly generating JavaScript

Reviewer: ```suggestion
advance, which can be helpful for prototyping or quickly generating JavaScript
```

object literals.

Melange provides some ways to create `Js.t` object values, as well as accessing
the properties inside them. To create values, the `[%bs.obj]` extension is used,
and the `##` infix operator allows to read from the object properties:

```ocaml
let john = [%bs.obj { name = "john"; age = 99 }]
let t = john##name
```

Which generates:

```js
var john = {
  name: "john",
  age: 99
};

var t = john.name;
```

> Note that type inference with objects is more lenient than with records. For

Reviewer: it's not "more lenient", it's structural (with optional row polymorphism) vs nominal typing.

I understand it might be harder to explain, but "more lenient" is a bit weird to say.

Author: I tweaked this part of the docs in https://github.com/melange-re/documentation-site/pull/12/commits/65a43aac373be89fafe069f10dc2926b94342126.

example, one can create a function that operates in all the object types that
include a field `name` that is of type string, e.g.:

```ocaml
let name_extended obj = obj##name ^ " wayne"

let one = name_extended [%bs.obj { name = "john"; age = 99 }]
let two = name_extended [%bs.obj { name = "jane"; address = "1 infinite loop" }]
```

> This is possible thanks to [objects

Reviewer: this is possible thanks to structural typing rather than polymorphism.

Author: I tweaked this section in https://github.com/melange-re/documentation-site/pull/12/commits/65a43aac373be89fafe069f10dc2926b94342126.

Reviewer: I don't know if we should mention subtyping, since we don't give examples for it. This is about structural typing vs nominal typing.

Author: removed refs to subtyping in https://github.com/melange-re/documentation-site/pull/12/commits/11740ebc41db0e86e170588969cc503af27b3736.

polymorphism](https://v2.ocaml.org/manual/objectexamples.html#s%3Apolymorphic-methods),
and is a characteristic shared by some types in OCaml, like objects and
polymorphic variants. The book Real World OCaml has [a good
introduction](https://dev.realworldocaml.org/objects.html) to object
polymorphism.

##### Bind to object properties

If you need to bind only to the property of a JavaScript object, you can use
`bs.get` and `bs.set`:

> ```ocaml

Reviewer: we should also include an example of what this codes generates, so that people can distinguish between `foo.bar` and `foo["bar"]`

type textarea
external set_name : textarea -> string -> unit = "name" [@@bs.set]
external get_name : textarea -> string = "name" [@@bs.get]
```

You can also use `bs.get_index` and `bs.set_index` to access a dynamic property
or an index:

```ocaml
type t
external create : int -> t = "Int32Array" [@@bs.new]
external get : t -> int -> int = "get" [@@bs.get_index]
external set : t -> int -> int -> unit = "set" [@@bs.set_index]

let () =
  let i32arr = (create 3) in
  set i32arr 0 42;
  Js.log (get i32arr 0)
```

Which generates:

```js
var i32arr = new Int32Array(3);
i32arr[0] = 42;
console.log(i32arr[0]);
```

#### Objects with dynamic shape (dictionary-like)

If you are binding to a JavaScript object that:

- might or might not add or remove keys
- contains only values that are of the same type

> Then it’s not really an object, it’s a dictionary or map. Use

Reviewer: let's say "then you can model it in OCaml using `Js.Dict`."

> "it's not really an object"

but, it is :P

Author: Rewrote this section fully in https://github.com/melange-re/documentation-site/pull/12/commits/019bf1a6c0d9429bdb97737f375f39471183cc80, please lmk what you think.

Reviewer: very nice!

of a JavaScript function that is called from Melange.

Let’s say we have the following JavaScript function, that we want to call from
Melange:

```js
// MyGame.js

function draw(x, y, border) {
  // let’s assume `border` is optional and defaults to false
}
draw(10, 20)
draw(20, 20, true)
```

When writing Melange bindings, we can add labeled arguments to make things more
clear:

```ocaml
external draw : x:int -> y:int -> ?border:bool -> unit -> unit = "draw"
  [@@module "MyGame"]

let () = draw ~x:10 ~y:20 ~border:true ()
let () = draw ~x:10 ~y:20 ()
```

Generates:

```js
var MyGame = require("MyGame");

MyGame.draw(10, 20, true);
MyGame.draw(10, 20, undefined);
```

The generated JavaScript function is the same, but now the usage in Melange is
much clearer.

**Note**: in this particular case, a final param of type unit, `()` must be
added after `border`, since `border` is an optional argument at the last
position. Not having the last param `unit` would lead to a warning, which is
explained in detail [in the OCaml
documentation](https://ocaml.org/docs/labels#warning-this-optional-argument-cannot-be-erased).

> Note that you can freely reorder the labels on the Melange side, they will

Reviewer: on the application side. but the definition side must match the order of arguments in JS. I think it's important to mention that.

Author: Fixed in https://github.com/melange-re/documentation-site/pull/12/commits/4d2c9ba756c3c28304e12ff8eacbc092a8ccd9ab.

appear in the right order on the JavaScript output:

```ocaml
external draw : x:int -> y:int -> ?border:bool -> unit -> unit = "draw"
  [@@module "MyGame"]
let () = draw ~x:10 ~y:20 ()
let () = draw ~y:20 ~x:10 ()
```

Generates:

```js
var MyGame = require("MyGame");

MyGame.draw(10, 20, undefined);
MyGame.draw(10, 20, undefined);
```

#### Calling a function as an object property

If we need to call a function that happens to be set as a JavaScript object
property, Melange provides the attribute `bs.send` so we can define it with just
one `external`:

```ocaml
(* Abstract type for the `document` global *)
type document

external document : document = "document" [@@bs.val]
external get_by_id : document -> string -> Dom.element = "getElementById"
  [@@bs.send]

let el = get_by_id document "my-id"
```

```js
var el = document.getElementById("my-id");
```

> When using `bs.send`, the first argument will be the object that holds the

Reviewer: should we also document `bs.send.pipe`, which inserts the argument at the end and is more suitable to be used with `|>`?

Author: Added in https://github.com/melange-re/documentation-site/pull/12/commits/80c512d7e8fa2571e645db5ef693d1acbc394585, lmk what you think as it implied quite a bit of changes.

I referred to the blog post I wrote back in the day, maybe we could include other links that explain OCaml pipe `|>`, not sure.

property with the function we want to call. In the example above, we are binding
to the JavaScript object `document` using an abstract type, a very useful
technique for bindings but also in other use cases. The "Encapsulation" section
[in the OCaml Cornell
textbook](https://cs3110.github.io/textbook/chapters/modules/encapsulation.html)
provides a great overview to understand what abstract types are and when they
are useful.

##### Chaining

It is common to find this kind of API in JavaScript: `foo().bar().baz()`. This
kind of API can be designed with Melange externals and the `bs.send` attribute,

> in combination with the [the pipe operator](todo-fix-me.md).

Reviewer: feels like an example is missing!

Author: Added in https://github.com/melange-re/documentation-site/pull/12/commits/7e661312d55c4a100e317f2ae7cf2baca071ca66

#### Variadic function arguments

Sometimes JavaScript functions take an arbitrary amount of arguments. For this
cases, Melange provides the `bs.variadic` attribute, which can be attached to
the `external` declaration. However, there is one caveat: all the variadic

> arguments need to belong to the same type.

Reviewer: should we document the solution for this?

```ocaml
type hide = Hide : 'a -> hide [@@unboxed]

external join : hide array -> string = "join" [@@bs.module "path"] [@@bs.splice]

let v = join [| Hide "a"; Hide 2 |]

```

which generates:

```js
// Generated by Melange
'use strict';

var Path = require("path");

var v = Path.join("a", 2);

exports.v = v;
/* v Not a pure module */
```

The same trick can be used for `Js.Dict`s

Author: Added solution for variadic in https://github.com/melange-re/documentation-site/pull/12/commits/b3bad3de602c738f17111b0e1689b0f3534ea247.

> The same trick can be used for `Js.Dict`s

The `unboxed` solution with existential types (is that what `hide` is?) for dictionaries would mean values can be added to the dictionary, but not be read from it, right?

Reviewer: yup, section looks good.

```ocaml
external join : string array -> string = "join"
  [@@bs.module "path"] [@@bs.splice]
let v = join [| "a"; "b" |]
```

Generates:

```js
var Path = require("path");
var v = Path.join("a", "b");
```

#### Bing to a polymorphic function

> Some JavaScript libraries will define functions that where the arguments can

Reviewer: ```suggestion
Some JavaScript libraries will define functions where the arguments can
```

> vary on both type and shape.There are two approaches to bind to those, depending

Reviewer: ```suggestion
vary on both type and shape. There are two approaches to bind to those, depending
```

If it is possible to enumerate the many forms an overloaded JavaScript function
can take, a flexible approach is to bind to each form individually:

```ocaml
external drawCat : unit -> unit = "draw" [@@bs.module "MyGame"]
external drawDog : giveName:string -> unit = "draw" [@@bs.module "MyGame"]
external draw : string -> useRandomAnimal:bool -> unit = "draw"
  [@@bs.module "MyGame"]
```

Note how all three externals bind to the same JavaScript function, `draw`.

##### Approach 2: Polymorphic variant + `bs.unwrap`

In some cases, the function has a constant number of arguments but the type of
the argument can vary. For cases like this, we can model the argument as a
variant and use the `bs.unwrap` attribute in the external.

Let’s say we want to bind to the following JavaScript function:

```js
function padLeft(value, padding) {
  if (typeof padding === "number") {
    return Array(padding + 1).join(" ") + value;
  }
  if (typeof padding === "string") {
    return padding + value;
  }
  throw new Error(`Expected string or number, got '${padding}'.`);
}
```

Here, `padding` can be modeled as a polymorphic variant.

```ocaml
external padLeft:
  string

>   -> ([ `Str of string

Reviewer: requirements:

- needs to be a polyvariant
- needs to contain payload

Author: Added in https://github.com/melange-re/documentation-site/pull/12/commits/11e3774e272a96fa359980d5fbaf4d76f7bd7a53, I also included the 2 other requisites.

      | `Int of int
      ] [@bs.unwrap])
  -> string
  = "padLeft" [@@bs.val]

let _ = padLeft "Hello World" (`Int 4)
let _ = padLeft "Hello World" (`Str "Message from Melange: ")
```

Which produces the following JavaScript:

```js
padLeft("Hello World", 4);
padLeft("Hello World", "Message from Melange: ");
```

As we saw in the [Non-shared data types](#non-shared-data-types) section, we
should rather avoid passing variants directly to the JavaScript side. By using
`bs.unwrap` we get the best of both worlds: from Melange we can use variants,
while JavaScript gets the raw values inside them.

#### Using polymorphic variants to bind to enums and string types

Some JavaScript APIs take a limited subset of values as input. For example,
Node’s `fs.readFileSync` second argument. It can only take a few given string
values: `"ascii"`, `"utf8"`, etc.

One could still bind it just as a string, but this would not prevent from using
unsupported values. Let’s see how we can use polymorphic variants and the
`bs.string` attribute to avoid runtime errors:

```ocaml
external read_file_sync :

>   name:string -> ([ `utf8 | `ascii ][@bs.string]) -> string = "readFileSync"

Reviewer: I think this is not needed anymore since polyvariants compile to strings.

Author: I tried removing `bs.string` but it seems it makes using `bs.as` impossible, tracked in https://github.com/melange-re/melange/issues/578. So for now decided to leave it.

Added an example with `bs.string` and `bs.as` in https://github.com/melange-re/documentation-site/pull/12/commits/15499fb1f3e577830b4fc6e3a4e8c29d28d51b15.

  [@@bs.module "fs"]

let _ = read_file_sync ~name:"xx.txt" `ascii
```

Which generates:

```js
var Fs = require("fs");
Fs.readFileSync("xx.txt", "ascii");
```

Attaching `bs.string` to the polymorphic variant type makes its constructor
compile to a string of the same name. This technique can be combined with the
`bs.as` attributes to modify the strings produced from the polymorphic variant
values, we will see an example of this just below.

Aside from producing string values, Melange also offers `bs.int` to produce
integer values:

```ocaml
external test_int_type :
  ([ `on_closed | `on_open [@bs.as 20] | `in_bin ][@bs.int]) -> int
  = "testIntType"
  [@@bs.val]

let value = test_int_type `on_open
```

In this example, `on_closed` will be encoded as 0, `on_open` will be 20 due to

> the attribute `bs.as` and `in_bin` will be 21.

Reviewer: we need to explain why: because it starts counting up from the previous value.

Author: Added in https://github.com/melange-re/documentation-site/pull/12/commits/15499fb1f3e577830b4fc6e3a4e8c29d28d51b15.

with the wildcard pattern `_`:

```ocaml
external process_on_exit : (_[@bs.as "exit"]) -> (int -> unit) -> unit
  = "process.on"
  [@@bs.val]

let () =
  process_on_exit (fun exit_code ->
    Js.log ("error code: " ^ string_of_int exit_code))
```

This generates:

```js
process.on("exit", function (exitCode) {
  console.log("error code: " + exitCode.toString());
});
```

The `bs.as "exit"` and the wildcard `_` pattern together will tell Melange to
compile the first argument of the JavaScript function to the string `"exit"`.

You can also use any JSON literal by passing a quoted string to `bs.as`: `bs.as
{json|true|json}` or `bs.as {json|{"name": "John"}|json}`.

#### Binding to callbacks

In OCaml, all functions have arity 1. This means that if you define a function
like this:

```ocaml
let add x y = x + y
```

Its type will be `int -> int -> int`. This means that one can partially apply
`add` by calling `add 1`, which will return another function expecting the
second argument of the addition. This kind of functions are called "curried"
functions, more information about currying in OCaml can be found in [this

> chapter](https://cs3110.github.io/textbook/chapters/hop/currying.html) of the

Reviewer: I wonder if we should compile all these external resources presented across the website and also expose them in a dedicated section.

Author: Maybe. Would this section be the "New to OCaml" section? Or some other? (Like "Melange for X users" in the JS section...)

Reviewer: Not a bad idea, and also something we can do anytime. no rush.

"OCaml Programming: Correct + Efficient + Beautiful" book.

> Currying is problematic when binding to JavaScript, let’s say we have a function

Reviewer: I don't like the word "problematic". We should probably just say "incompatible calling conventions", no?

Author: Removed "problematic" and expanded a bit in https://github.com/melange-re/documentation-site/pull/12/commits/127995137166ab4cc9c597b6ff9266fa2c7865e2.

  [@@bs.val]
```

Unfortunately, this is not completely correct. The issue is in the callback
function, with type `'a → 'b → 'c`. This means that `map` will expect a function
like `add` described above. But as we said, in OCaml, having two arguments means
just to have two functions that take one argument.

Let’s rewrite `add` to make the problem a bit more clear:

```ocaml
let add x = let partial y = x + y in partial
```

This will be compiled to:

```javascript
function add(x) {
  return (function (y) {
    return x + y | 0;
  });
}
```

Now if we ever used our external function `map` with our `add` function by
calling `map arr1 arr2 add` it would not work as expected. JavaScript function
application does not work the same as in OCaml, so the function call in the
`map` implementation, `f(a[i],b[i])`, would be applied over the outer JavaScript
function `add`, which only takes one argument `x`, and `b[i]` would be just
discarded. The value returned from the operation would not be the addition of
the two numbers, but rather the inner anonymous callback.

To solve this mismatch between OCaml and JavaScript functions and their

> application, Melange provides a special attribute `@bs` that can be used with

Reviewer: should we also mention the ReasonML syntax sugar? `f(. a, b)`.

Author: All the docs should be revisited with Reason syntax in mind. Some other things that comes to mind are:
- Pipe operator (`|.` vs `->`)
- Some syntax like extensions differ (e.g. two percentage signs in `ml` are one in `re`)

I would rather tackle all this as a focused task. Added some additional note in #7: https://github.com/melange-re/documentation-site/issues/7#issuecomment-1549519557

external functions.

In the example above:

```ocaml
external map : 'a array -> 'b array -> (('a -> 'b -> 'c)[@bs]) -> 'c array
  = "map"
  [@@bs.val]
```

> Here `('a → 'b → 'c [@bs])` will be interpreted as having arity 2, in general,

Reviewer: what's this `→` character? can we use `->` instead?

Author: Removed in https://github.com/melange-re/documentation-site/pull/12/commits/ce220da061bc231ee5c1cb4f6ab96a2ce0167746 (also replaced the `…​` weird char with `...`)

correct to declare `x.onload` of type `unit -> unit`. Instead, Melange
introduces a special attribute, `bs.this`, which allows to type `x` as this:

```ocaml
type x
external x : x = "x" [@@bs.val]
external set_onload : x -> ((x -> int -> unit)[@bs.this]) -> unit = "onload"
  [@@bs.set]
external resp : x -> int = "response" [@@bs.get]
let _ =
  set_onload x
    begin
      fun [@bs.this] o v -> Js.log (resp o + v)
    end
```

Which generates:

```javascript
x.onload = function (v) {
  var o = this;
  console.log((o.response + v) | 0);
};
```

Note that the first argument will be reserved for `this`.

#### Wrapping returned nullable values

For JavaScript functions that return a value that can be `undefined` or `null`,
Melange provides `bs.return`. Using this attribute will have Melange generated
code automatically convert the value returned by the function to an `option`
type that can be used safely from Melange side, avoiding the need to use manual
conversion functions like `Js.Nullable.toOption` and such.

```ocaml
type element
type document
external get_by_id : document -> string -> element option = "getElementById"
  [@@bs.send] [@@bs.return nullable]

let test document =

>   let elem = document |. get_by_id "header" in

Reviewer: have we introduced `|.` before? This will be unfamiliar to people if we haven't.

Author: There is a "Chaining" section before this one that talks about it, but I agree better to remove it to avoid distractions. Done in https://github.com/melange-re/documentation-site/pull/12/commits/ce220da061bc231ee5c1cb4f6ab96a2ce0167746.

  match elem with
  | None -> 1
  | Some _element -> 2
```

Which generates:

```js
function test($$document) {
  var elem = $$document.getElementById("header");
  if (elem == null) {
    return 1;
  } else {
    return 2;
  }
}
```

The `bs.return` attribute takes an attribute payload, as seen with `[@@bs.return
nullable]` above. Currently 4 directives are supported: `null_to_opt`,
`undefined_to_opt`, `nullable` and `identity`.

`nullable` is encouraged, as it will convert from `null` and `undefined` to
`option` type.

<!-- When the return type is unit: the compiler will append its return value with an OCaml unit literal to make sure it does return unit. Its main purpose is to make the user consume FFI in idiomatic OCaml code, the cost is very very small and the compiler will do smart optimizations to remove it when the returned value is not used (mostly likely). -->

`identity` will make sure that compiler will do nothing about the returned

> value. It is rarely used, but introduced here for debugging purpose.

Reviewer: ```suggestion
value. It is rarely used, but introduced here for debugging purposes.
```
